File: frontend/fundrbolt-admin/src/services/event-dashboard.ts
import apiClient from '@/lib/axios'

export type ScenarioType = 'base' | 'optimistic' | 'conservative'
export type SegmentType = 'table' | 'guest' | 'registrant' | 'company'

export interface MoneyValue {
  amount: number
  currency: string
}

export interface RevenueSourceSummary {
  source: string
  actual: MoneyValue
  projected: MoneyValue
  target: MoneyValue
  variance_amount: MoneyValue
  variance_percent: number
  pacing_percent: number
}

export interface PacingStatus {
  status: 'on_track' | 'off_track'
  pacing_percent: number
  trajectory: 'linear'
}

export interface CashflowPoint {
  date: string
  actual: MoneyValue
  projected: MoneyValue
}

export interface WaterfallStep {
  label: string
  amount: MoneyValue
}

export interface FunnelStage {
  stage: 'invited' | 'registered' | 'checked_in' | 'donated_bid'
  count: number
}

export interface AlertCard {
  source: string
  status: 'active' | 'resolved'
  threshold_percent: number
  consecutive_refreshes: number
  triggered_at?: string | null
}

export interface DashboardSummary {
  event_id: string
  goal: MoneyValue
  total_actual: MoneyValue
  total_projected: MoneyValue
  variance_amount: MoneyValue
  variance_percent: number
  pacing: PacingStatus
  sources: RevenueSourceSummary[]
  waterfall: WaterfallStep[]
  cashflow: CashflowPoint[]
  funnel: FunnelStage[]
  alerts: AlertCard[]
  last_refreshed_at: string
}

export interface SegmentBreakdownItem {
  segment_id: string
  segment_label: string
  total_amount: MoneyValue
  contribution_share: number
  guest_count: number
}

export interface SegmentBreakdownResponse {
  segment_type: SegmentType
  items: SegmentBreakdownItem[]
}

export interface ProjectionAdjustment {
  source: string
  projected: MoneyValue
}

export interface ProjectionAdjustmentSet {
  event_id: string
  scenario: ScenarioType
  adjustments: ProjectionAdjustment[]
  updated_at: string
  updated_by: string
}

export interface ProjectionAdjustmentUpdate {
  scenario: ScenarioType
  adjustments: ProjectionAdjustment[]
}

class EventDashboardService {
  async getDashboard(eventId: string, scenario: ScenarioType = 'base') {
    const response = await apiClient.get<DashboardSummary>(
      `/admin/events/${eventId}/dashboard`,
      {
        params: { scenario },
      }
    )
    return response.data
  }

  async getProjections(eventId: string, scenario: ScenarioType = 'base') {
    const response = await apiClient.get<ProjectionAdjustmentSet>(
      `/admin/events/${eventId}/dashboard/projections`,
      {
        params: { scenario },
      }
    )
    return response.data
  }

  async updateProjections(eventId: string, payload: ProjectionAdjustmentUpdate) {
    const response = await apiClient.post<ProjectionAdjustmentSet>(
      `/admin/events/${eventId}/dashboard/projections`,
      payload
    )
    return response.data
  }

  async getSegments(
    eventId: string,
    segmentType: SegmentType,
    limit = 20,
    sort: 'total_amount' | 'contribution_share' = 'total_amount'
  ) {
    const response = await apiClient.get<SegmentBreakdownResponse>(
      `/admin/events/${eventId}/dashboard/segments`,
      {
        params: {
          segment_type: segmentType,
          limit,
          sort,
        },
      }
    )
    return response.data
  }
}

export const eventDashboardService = new EventDashboardService()

File: backend/app/schemas/event_dashboard.py
"""Pydantic schemas for admin event dashboard endpoints."""

from __future__ import annotations

from datetime import date, datetime
from typing import Literal
from uuid import UUID

from pydantic import BaseModel, Field

ScenarioType = Literal["base", "optimistic", "conservative"]
SegmentType = Literal["table", "guest", "registrant", "company"]
SortType = Literal["total_amount", "contribution_share"]


class MoneyValue(BaseModel):
    """Monetary value representation."""

    amount: float = Field(default=0)
    currency: str = Field(default="USD")


class RevenueSourceSummary(BaseModel):
    """Summary for one revenue source."""

    source: str
    actual: MoneyValue
    projected: MoneyValue
    target: MoneyValue
    variance_amount: MoneyValue
    variance_percent: float = Field(default=0)
    pacing_percent: float = Field(default=0)


class PacingStatus(BaseModel):
    """Pacing status for an event."""

    status: Literal["on_track", "off_track"]
    pacing_percent: float = Field(default=0)
    trajectory: Literal["linear"] = "linear"


class CashflowPoint(BaseModel):
    """Timeline point for cashflow chart."""

    date: date
    actual: MoneyValue
    projected: MoneyValue


class WaterfallStep(BaseModel):
    """Waterfall chart step."""

    label: str
    amount: MoneyValue


class FunnelStage(BaseModel):
    """Conversion funnel stage."""

    stage: Literal["invited", "registered", "checked_in", "donated_bid"]
    count: int = Field(default=0, ge=0)


class AlertCard(BaseModel):
    """Dashboard alert item."""

    source: str
    status: Literal["active", "resolved"]
    threshold_percent: float = Field(default=90)
    consecutive_refreshes: int = Field(default=0, ge=0)
    triggered_at: datetime | None = None


class DashboardSummary(BaseModel):
    """Main dashboard payload."""

    event_id: UUID
    goal: MoneyValue
    total_actual: MoneyValue
    total_projected: MoneyValue
    variance_amount: MoneyValue
    variance_percent: float = Field(default=0)
    pacing: PacingStatus
    sources: list[RevenueSourceSummary]
    waterfall: list[WaterfallStep]
    cashflow: list[CashflowPoint]
    funnel: list[FunnelStage]
    alerts: list[AlertCard]
    last_refreshed_at: datetime


class SegmentBreakdownItem(BaseModel):
    """Breakdown row for a segment."""

    segment_id: str
    segment_label: str
    total_amount: MoneyValue
    contribution_share: float = Field(default=0)
    guest_count: int = Field(default=0, ge=0)


class SegmentBreakdownResponse(BaseModel):
    """Segment breakdown payload."""

    segment_type: SegmentType
    items: list[SegmentBreakdownItem]


class ProjectionAdjustment(BaseModel):
    """Projected value for a source."""

    source: str
    projected: MoneyValue


class ProjectionAdjustmentSet(BaseModel):
    """Projection scenario values."""

    event_id: UUID
    scenario: ScenarioType
    adjustments: list[ProjectionAdjustment]
    updated_at: datetime
    updated_by: str


class ProjectionAdjustmentUpdate(BaseModel):
    """Request body for projection updates."""

    scenario: ScenarioType = "base"
    adjustments: list[ProjectionAdjustment] = Field(default_factory=list)
